### PR TITLE
Fix handling of initial columns in CrossFilter

### PR DESCRIPTION
**Pull request by rothnic**
This was intended to be a fix for the barchart implementation in crossfilter, based on [this comment thread](https://groups.google.com/a/continuum.io/forum/?utm_medium=email&utm_source=footer#!msg/bokeh/2StCzFnKj_0/lTb2V8Kc8AEJ), but I don't think there is anything directly wrong with it.

I took the provided data set for cruises and found that the issue is really in classifying the data so that it is handled correctly in crossfilter. As of right now, cf doesn't try to interpret what kind of data it actually is, independent of its current form. In this case, a column of years as integers causes cf to treat the data as continuous, which results in a binning of the column so that it can be plotted as a bar chart/pseudo histogram.

After converting the column to a string, the bar chart was produced as expected. However, I found that there was an issue with handling a dataset that contains fewer than 2 continuous columns. The first commit handles this issue.

The second commit addresses the following items to make the bar chart plugin more robust:
* Add count aggregation type, and 'None' selection to the columns. This makes it possible to produce simple bar charts of just counting the unique things in the x column.
* Make sure that column data source creation for bar chart can occur with and without a y selection. 
* Make the x axis of the plot rotate if it is longer than a certain length.
* Handle more "bad" configurations of bar chart and notify the user of the issue via the plot title.

I still need to do some more regression testing to make sure I didn't break other plugins, and to test against other columns. I'll let everyone know when complete.

### CODE REVIEW

**Comment by bryevdv** [comment]
@rothnic thanks for this work!

**Comment by rothnic** [comment]
No problem. Here is a gallery of some of the different configurations and what the result is: [gallery of cruise data](http://imgur.com/a/md4Xf)

**Comment by rothnic** [comment]
Added a percent aggregation type. This should probably be broken out as a histogram plugin, but the concepts are fairly similar for this simple implementation. There is no grouping and plotting of multiple distributions in one plot, but the faceting can handle this:

image

**Comment by fabioannovazzi** [comment]
This is great Nick!

I also added a dummy "count" column and set the year to str, and my columns
now space nicely in the graph. I am not familiar with github and pull
requests - when will your new functionality be added to the main thing (I
installed 8.2 and they are not there). Or otherwise is there a simple way
of working with still not merged updates of Bokeh. I know this is a stupid
question, but I really have not idea how to di it/if it is a good idea to
do it.

Cheers

Fabio

Fabio Annovazzi
tel. 0041 (0)78-6334750
www.logeeka.com

**Comment by fpliger** [comment]
@fabioannovazzi This PR was not ready to be merged before the 0.8.2. (btw, @rothnic I've set it to `WIP` but if you think it's ready for a review let us know..).

You can start playing with this branch by cloning this branch. Otherwise you can:

- wait until it's merged to the master branch and keep updated to that (you can have more details about the process to do this by looking at the [dev guide](http://bokeh.pydata.org/en/latest/docs/dev_guide.html) and specifically the [building part](http://bokeh.pydata.org/en/latest/docs/dev_guide/building.html). I don't think this PR will wait too long before landing to master, but @rothnic surely have a better understanding of that.

- wait for the next dev build after this PR is merged to master. We release dev builds quite often so it's likely that you won't wait too much.

- wait for the next major release. This is the option that will take longer as we just released 0.8.2 :-)

If your scope and purposes is more exploratory about the features and code I'd suggest you to try and play a bit with git and branches. This will enable you to use whatever branch you want and is always a good exercise.  :-)

**Comment by rothnic** [comment]
Caught one little bug, but this should be good to go. I ran through each chart type with a bunch of configurations. Really need to look into automating this...

As mentioned before, I'm looking to integrate charts in at some point, but this extends the bar chart out a bit to do some histogram-like functionality for now via a count or percent aggregation type. There is some logic in there to catch the configurations that would break the plot, so it notifies the user as to how they can fix the configuration.

File: bokeh/crossfilter/models.py [hunk 1]
```diff
@@ -282,7 +285,14 @@ def make_plot_choices(self):
           dict: x, y, and plot_type keys and values for initial setup
 
         """
-        x, y = [x['name'] for x in self.continuous_columns()[:2]]
+        cont_columns = self.continuous_columns
+
+        # prefer continuous columns to initialize with, otherwise use what we have
+        if len(cont_columns) > 1:
+            x, y = [x['name'] for x in self.continuous_columns[:2]]
```
Comment (fpliger) [inline bokeh/crossfilter/models.py]: To be consistent, `cont_columns` or use `self.continuous_columns` for the `len` check

File: bokeh/crossfilter/plotting.py [hunk 2]
```diff
@@ -93,26 +92,43 @@ def make_continuous_bar_source(df, x_field, y_field, agg):
...
     # Convert back to a DataFrame on the aggregated data
-    result = aggregate().reset_index()
+    if y_field == 'None':
+        if agg == 'count':
+            series = aggregate()
+
+        result = pd.DataFrame(
+            data={x_field: series.index, agg: series.values})
+    else:
+        result = aggregate().reset_index()
```
Comment (fpliger) [inline bokeh/crossfilter/plotting.py]: What happens if `agg == "percent"' here?

**Comment by fpliger** [comment]
Left a couple of comments. 

@rothnic FYI, I really hope to find out some time to work on the crossfilter and extend it. I like the idea of integrating charts and actually letting users extend crossfilter with custom plugins in general. Need to better study the current implementation. :-)

**Comment by rothnic** [comment]
@fpliger thanks, i reorganized the categorical bar chart changes to be a bit more explicit for separating the x only and x/y aggregations.

**Comment by rothnic** [comment]
This is cleaned up to pass the flake8 failures it was encountering and has had the latest master merged into it.

**Comment by bryevdv** [comment]
@rothnic just to be clear this should move from `WIP` -> `ready`?

**Comment by rothnic** [comment]
@bryevdv that is what I was thinking, but after talking to @fabioannovazzi, who is using it, I'm going to make some modifications

**Comment by rothnic** [comment]
This one is ready now.

**Comment by bryevdv** [comment]
LGTM

**Pull request merged**